measurements/pages/__init__.py

```diff
@@ -204,34 +208,69 @@ def files_in_country(country_code):
-        # XXX
-        # We have duplicate measurements :(
-        # So the below exception actually happens. This should be resolved
-        # in the data processing pipeline.
-        #   Report.filename == filename).one()
-    except NoResultFound:
-        raise NotFound("No file with that filename found")
-    except MultipleResultsFound:
-        # This should never happen.
-        raise HTTPException("Duplicate measurement detected")
-
-    if report is None:
+def decompress_autoclaved(autoclaved_filename, frames_off, frames_size, intra_off, intra_size):
+    def generator():
+        url = urljoin(current_app.config['AUTOCLAVED_BASE_URL'], autoclaved_filename)
+
+        current_app.logger.debug("Fetching autoclaved from: %s" % url)
+        current_app.logger.debug("  frames_off: %d" % frames_off)
+        current_app.logger.debug("  frames_size: %d" % frames_size)
+        current_app.logger.debug("  intra_off: %d" % intra_off)
+        current_app.logger.debug("  intra_size: %d" % intra_size)
+        try:
+            r = requests.get(url, headers={"Range": "bytes={}-{}".format(frames_off, frames_off + frames_size)},
+                                stream=True)
+            streamed_data = 0
+            for chunk in lz4framed.Decompressor(r.raw):
+                current_app.logger.debug("chunking away 1")
+                d = chunk
+                current_app.logger.debug("chunking away 2")
+                if streamed_data == 0 and intra_off > 0:
+                    d = chunk[intra_off:]
+                if streamed_data > (intra_size + len(d)):
+                    d = chunk[:(streamed_data - intra_size)]
+                yield d
+                streamed_data += len(d)
+            current_app.logger.debug("sent: %d" % streamed_data)
+        except Exception as exc:
+            current_app.logger.error("failed to fetch streamed data: %s" % exc)
+            raise HTTPException("Failed to fetch data")
+    return generator
+
+@pages_blueprint.route('/files/download/<path:textname>')
+def files_download(textname):
+    q = current_app.db_session.query(
+            Measurement.id.label('m_id'),
+            Measurement.report_no.label('report_no'),
+            Measurement.frame_off.label('frame_off'),
+            Measurement.frame_size.label('frame_size'),
+            Measurement.intra_off.label('intra_off'),
+            Measurement.intra_size.label('intra_size'),
+            Report.report_no.label('r_report_no'),
+            Report.autoclaved_no.label('r_autoclaved_no'),
+            Autoclaved.filename.label('a_filename'),
+            Autoclaved.autoclaved_no.label('a_autoclaved_no'),
+    ).filter(Report.textname == textname) \
+        .join(Report, Report.report_no == Measurement.report_no) \
+        .join(Autoclaved, Autoclaved.autoclaved_no == Report.autoclaved_no) \
+        .order_by(Measurement.frame_off.asc(), Measurement.intra_off.asc())
```

Comment:
```SQL
SELECT Measurement.id, Measurement.report_no, frame_off, frame_size, intra_off, intra_size, Report.report_no, Report.autoclaved_no, Autoclaved.filename, Autoclaved.autoclaved_no FROM Measurement, Report, Autoclaved WHERE Report.report_no == Measurement.report_no AND Autoclaved.autoclaved_no == Report.autoclaved_no ORDER BY Measurement.frame_off, Measurement.intra_ff ASC```?

Reply:
Kind of, except you are missing the two JOINs on the Report and Autoclaved tables.

Reply:
Okay, good!